src/estimators/stratified_sketch.py

```diff
@@ -0,0 +1,261 @@
+    # (3) a 1+ reach sketch.
+    self.sketches = {}
+    self.seed = random_seed
+    self.max_freq = max_freq
+
+  @classmethod
+  def init_from_exact_multi_set(cls, max_freq, exact_multi_set,
+                                cardinality_sketch_factory, random_seed):
+    """Initialize a Stratified sketch from one ExactMultiSet.
+
+    Args:
+      exact_multi_set: ExactMultiSet object to use for initialization.
+      CardinalitySketch: Class type of cardinality sketches this stratified
+        sketch will hold.
+    """
+    assert (cardinality_sketch_factory is
+            not None), ('cardinality_sketch is None')
+    stratified_sketch = cls(
+        max_freq=max_freq,
+        cardinality_sketch_factory=[],
+        random_seed=random_seed)
+    stratified_sketch.sketches[ONE_PLUS] = cardinality_sketch_factory(
+        random_seed)
+    stratified_sketch.sketches[ONE_PLUS].add_ids(
+        list(exact_multi_set.ids().keys()))
+
+    reversedict = {}
+    for k, v in exact_multi_set.ids().items():
+      reversedict.setdefault(min(v, max_freq), []).append(k)
+
+    for freq in range(1, max_freq + 1):
+      stratified_sketch.sketches[freq] = cardinality_sketch_factory(random_seed)
+      if (freq in reversedict):
+        stratified_sketch.sketches[freq].add_ids(reversedict[freq])
+
+    return stratified_sketch
+
+  @classmethod
+  def init_from_set_generator(cls, max_freq, set_generator,
+                              cardinality_sketch_factory, random_seed):
+    """Initialize a Stratified sketch from a Set Generator.
+
+    Args:
+      set_generator: SetGenerator object to draw ids from for initialization.
+      CardinalitySketch: Class type of cardinality sketches this stratified
+        sketch will hold.
+    """
+    assert (cardinality_sketch_factory is
+            not None), ('cardinality_sketch is None')
+    exact_multi_set = ExactMultiSet()
+    for generated_set in set_generator:
+      exact_multi_set.add_ids(generated_set)
+    return cls.init_from_exact_multi_set(max_freq, exact_multi_set,
+                                         cardinality_sketch_factory,
+                                         random_seed)
+
+  def assert_compatible(self, other):
+    """"Check if the two StratifiedSketch are comparable.
+
+    Args:
+      other: the other StratifiedSketch for comparison.
+
+    Raises:
+      AssertionError: if the other sketches are not StratifiedSketch, or if
+      their random_seed are different, or if the frequency targets are
+      different.
+    """
+    assert isinstance(other,
+                      StratifiedSketch), ('other is not a StratifiedSketch.')
+    assert self.seed == other.seed, ('The random seeds are not the same: '
+                                     f'{self.seed} != {other.seed}')
+    assert self.max_freq == other.max_freq, (
+        'The frequency targets are different: '
+        f'{self.max_freq} != {other.max_freq}')
+
+
+class PairwiseEstimator(object):
```

Comment:
A little bit confused with this name. Will this be used as estimator for this frequency-based bloom filter? We probably need to inherit `EstimatorBase` and need to return a list in __call__?

Reply:
Thanks for pointing this out. Yes, this should be the case. I was thinking of adding the estimation in a following PR but this seems like a good place to do it. I added the estimation as well.